__noinline__ the top level igamma cuda kernel.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47414 __noinline__ the top level igamma cuda kernel.**
* #47410 Move igamma cuda specific code to kernel file.

This improves the build time of this file by 10x on my machine (~12 minutes to ~1 minute).

Differential Revision: [D24746458](https://our.internmc.facebook.com/intern/diff/D24746458)

gchanan committed:
__noinline__ the top level igamma cuda kernel.

This improves the build time of this file by 10x on my machine (~12 minutes to ~1 minute).

gchanan commented:
Thanks to @mruberry and @ngimel for help with the fix.

mruberry commented on aten/src/ATen/native/cuda/IGammaKernel.cu:
```diff
@@ -365,7 +365,7 @@ __host__ __device__ scalar_t _igamc_helper_continued_fraction(scalar_t a, scalar
 
 template <typename scalar_t>
-inline __host__ __device__ scalar_t calc_igammac(scalar_t a, scalar_t x) {
+__host__ __device__ scalar_t calc_igammac(scalar_t a, scalar_t x) {
```
Do you want to preemptively `__noinline__` this, too? It's about to be exposed directly to users.

gchanan replied:
will do

mruberry reviewed (approved):
Great detective work!

gchanan committed:
Update on "__noinline__ the top level igamma cuda kernel."


This improves the build time of this file by 10x on my machine (~12 minutes to ~1 minute).

Differential Revision: [D24746458](https://our.internmc.facebook.com/intern/diff/D24746458)

gchanan committed:
Update on "__noinline__ the top level igamma cuda kernel."


This improves the build time of this file by 10x on my machine (~12 minutes to ~1 minute).

Differential Revision: [D24746458](https://our.internmc.facebook.com/intern/diff/D24746458)
